bevy_lint/src/paths.rs

```diff
@@ -1,58 +1,74 @@
-//! A collection of hardcoded types and functions that the linter uses.
+//! A collection of hardcoded type and function paths that the linter uses.
 //!
 //! Since Bevy is a 3rd-party crate, we cannot easily add diagnostic items to it. In lieu of this,
 //! we hardcode the paths to the items we need here, for easy referencing.
-//!
-//! Also see: [`match_type()`](clippy_utils::ty::match_type),
-//! [`match_def_path()`](clippy_utils::match_def_path).
 
-/// <https://github.com/bevyengine/bevy/blob/release-0.16.0/crates/bevy_app/src/app.rs#L78>
-pub const APP: [&str; 3] = ["bevy_app", "app", "App"];
+use clippy_utils::paths::{PathLookup, PathNS};
+
+use crate::sym;
+
+/// Returns a new [`PathLookup`] in the [type namespace](PathNS::Type) for a given path.
+///
+/// `type_path!()` takes a `::`-separated list of identifiers. Each identifier should correspond to
+/// a [`Symbol`](rustc_span::Symbol) in [`crate::sym`]. For example,
+/// `type_path!(bevy_app::app::App)` creates a [`PathLookup`] in the [`PathNS::Type`] namespace
+/// with the path `[sym::bevy_app, sym::app, sym::App]`.
+macro_rules! type_path {
+    ($first:ident $(:: $remaining:ident)*) => {
+        PathLookup::new(PathNS::Type, &[sym::$first, $(sym::$remaining),*])
+    };
+}
```

Comment:
This macro is specifically for the type namespace. If we end up defining paths for the value or macro namespaces, we'll need to copy this macro. (See [`PathNS`](https://docs.rs/clippy_utils/0.1.89/clippy_utils/paths/enum.PathNS.html) for more details.)